2.3-TapTree.ipynb

```diff
@@ -0,0 +1,500 @@
+    "\n",
+    "\n",
+    "## Part 1: TapTree Construction.\n",
+    "\n",
+    "### 1A - TapTree Descriptors\n",
+    "\n",
+    "Descriptors are a high-level template language to describe individual or ranges of outputs. \n",
+    "\n",
+    "* Human readable & descriptive\n",
+    "* Imply standard script patterns\n",
+    "* Composable.\n",
+    "\n",
+    "For taproot, we currently propose a taproot descriptor expression with the following pattern. \n",
+    "\n",
+    "![test](images/TapTree0.jpg)\n",
+    "\n",
+    "The tree structure is implied by the nested expression:\n",
+    "* **An internal node is represented by its children:**\n",
+    "    * Node expression: \n",
+    "        * `[child_left, child_right]`\n",
+    "    * Node expressions are composable: \n",
+    "        * `[child_left, [child_left', child_right']]`\n",
+    "* **A leaf node can be represented by its TapScript descriptor:**\n",
+    "    * For example, a pay-to-pubkey TapScript: \n",
+    "        * `ts(pk(key))`\n",
+    "    * _More on TapScript descriptors in chapter 2.4._\n",
+    "    \n",
+    "*Note: The proposed taproot descriptor is not unique for a given taproot output. A given taproot output may have multiple correct taproot descriptor expressions. This is because we do not impose lexicographic ordering of the descriptor leafs and nodes as is done when computing the taproot.*"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "#### Example 1.1 - Constructing a Taptree from a descriptor.\n",
+    "\n",
+    "![test](images/TapTree1.jpg)\n",
+    "\n",
+    "In this example, we will construct the TapTree shown above from a descriptor literal.\n",
+    "\n",
+    "* **Class: `TapTree`**\n",
+    "    * Construct from descriptor:\n",
+    "        * `TapTree.from_desc(descriptor_string)`\n",
+    "    * Serialize back to descriptor:\n",
+    "        * `TapTree.desc`"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "# Generate internal keypairs\n",
+    "privkey_internal = ECKey()\n",
+    "privkey_internal.generate()\n",
+    "pubkey_internal = privkey_internal.get_pubkey()\n",
+    "\n",
+    "# Generaet keypairs for the tapleaf pay-to-pubkey scripts.\n",
+    "privkeyA = ECKey()\n",
+    "privkeyB = ECKey()\n",
+    "privkeyC = ECKey()\n",
+    "privkeyA.generate()\n",
+    "privkeyB.generate()\n",
+    "privkeyC.generate()\n",
+    "pubkeyA = privkeyA.get_pubkey()\n",
+    "pubkeyB = privkeyB.get_pubkey()\n",
+    "pubkeyC = privkeyC.get_pubkey()\n",
+    "\n",
+    "# Construct Descriptor String.\n",
+    "ts_desc_A = 'ts(pk({}))'.format(pubkeyA.get_bytes().hex())\n",
+    "ts_desc_B = 'ts(pk({}))'.format(pubkeyB.get_bytes().hex())\n",
+    "ts_desc_C = 'ts(pk({}))'.format(pubkeyC.get_bytes().hex())\n",
+    "tp_desc = 'tp({},[[{},{}],{}])'.format(pubkey_internal.get_bytes().hex(),\n",
```

Comment:
done. I think this could be made even clearer by using f-strings (as long as we can specify that users must be running Python 3.6 or higher. This line becomes:

`tp_desc = f'tp({pk_hex},[[{ts_desc_A},{ts_desc_B}],{ts_desc_C}])'`